### DISCUSSION TITLE
How to use uvToRenderTargetUV() for both renderTarget or imported input

### DISCUSSION THREAD

**Discussion by kjw0723** [google/filament#6363] [Q&A] [answered]
Hello,

First of all, thanks for making this amazing rendering library.

I have a question about using `uvToRenderTargetUV()`.

In our use case, we need to support both texture types(renderTarget or imported) in our material at runtime.
In order to support that, my guess is that we should write a bool parameter to check if input texture is renderTarget or not.
This approach would seem work, but we might have to write this parameter for all our materials.

Another one is making renderTargets for all imported textures and always use `uvToRenderTargetUV()`.

I would like to know if there is a better method to support both types of texture input in one single material source.

Any suggestions about this? I would appreciate your help.
Thank you for your time.

**Comment by romainguy** [comment]
@pixelflinger would be the best person to answer this question. This only matters if you target Vulkan or Metal btw. OpenGL backends don't do anything in `uvToRenderTargetUV()`.

**Comment by pixelflinger** [comment]
Hi, I think I need a little bit more context. I don't quite understand what do you mean by "we need to support both texture types(renderTarget or imported)".  
- what do you mean by "support"
- what's different between both scenario (i.e. what's not working the same)?
- how do you import the texture?
- which backends and platform are you targeting? Is this specific to Android or no?

**Reply by kjw0723** [reply]
@pixelflinger Hello,

- We are making a RenderTarget chaining system for image filtering like [this page](https://developer.apple.com/documentation/coreimage/processing_an_image_using_built-in_filters?language=objc). So if we make an image filter the source texture would be either `setImage()` or RenderTarget attachment.
- According to [material guide](https://google.github.io/filament/Materials.html), if you sample a texture attached to a RenderTarget, you need to use `uvToRenderTargetUV()` or `getUV0()`. So as we don't know which texture would be provided to a material at runtime, we thought there should be a bool variable in a material to determine which sampling function to use
- We create a texture form an image (using `setImage()`) or RenderTarget attachment.
- We are using OpenGL for Android and Metal for iOS.

**Comment by pixelflinger** [comment, marked as answer]
Okay so let me just start by saying that all this is a real mess and some of it is a bit self-inflicted. I think I need to explain the issues before trying to give a direct answer to your question.

**Open GL**

First, let's ignore filament for a moment, and just consider OpenGL. In GL, when you upload texture data from the CPU (e.g. with `glTexImage2D`), the low addresses in your CPU buffer represent the low `v` texture coordinates. In other words,  if you pass a "normal" image, such as a jpeg or png to `glTexImage2D`, the **top** of the image is a `v == 0`. This sounds reasonable at this point, but read on...

Now, if you use that texture as a color attachment to a FBO, the origin is at the left-**bottom**, so the `v == 0` coordinate is at the **bottom**. This means that your texture appears vertically flipped on screen.

If you then use `glReadPixels()` to copy the framebuffer content to a CPU buffer, the **bottom** of the screen is going to be transferred to the low addresses of your CPU buffer, and as a result, the image it going to be "unflipped" and look as it should, on the CPU side.

All of this is self-consistant, it's just that the traditional CPU side "up" is different from what GL expects, but things are inverted twice when you use both `glTexImage2D` and `glReadPixels` -- however in "the middle" the image in the color buffer is flipped.

The correct solution to this is actually to flip your "traditional" CPU side images vertically before calling `glTexImage2D` and flip again after calling `glReadPixels`.

**Metal and Vulkan**

In Metal and Vulkan, the origin of both textures and attachments is left-top, so that "flipping" doesn't happen. In other words, in Metal and Vulkan you don't need to vertically flip your CPU data respectively before uploading to the GPU and after downloading from it, for it to appear as expected on screen.

**Sampling from textures**

If you just sample from your textures, no inversion occurs in all APIs:
- In OpenGL the low-addresses of the CPU buffer correspond to the low values of `v`
- In Metal and Vulkan the low-addresses of the CPU buffer also correspond to the low values of `v`
So, *uploading* data to textures and sampling it is not a problem.

... well, unless you followed my advice above and inverted your data before sending it. oh boy...

**Drawing into attachments**

Note that at this point we haven't drawn anything, we've just loaded some CPU data into a texture and made that texture a color attachment, then we read back from that framebuffer.

When drawing geometry, what matters is the *clip-space*. In both OpenGL and Metal, the clip-space is *y-up*, in Vulkan it's *y-down*. Filament always uses the OpenGL convention and inverts the Vulkan clip-space y coordinate in the vertex shader.

But now we have a problem with Metal; Metal's clip-space is *y-up* but its window coordinate system is *y-down*! As mentioned before the origin is at *left-top*. So... in Metal, if you draw a pyramid in clip-space (meaning the tip is at high y coordinates) that corresponds to **low** `v` texture coordinates (!!!).  

In OpenGL, clip-space is *y-up* and window coordinates are *y-up* too, so no inversion there.

In Vulkan, clip-space is *y-down* and window coordinates are also *y-down*, so no inversion either, **but** remember that we are always using the *OpenGL* clip-space convention in Filament, so our situation is similar to Metal and we do have an inversion.


**Filament**

Filament attempts to fix these issues but does so only partially.

First we are addressing the 2nd problem where *clip-space* and *window-coordinates* don't agree. This is done with `uvToRenderTargetUV()`, this is why this method exists. So you should **always** use it when sampling from a texture that had geometry rendered into it, even if that geometry is just a quad.

We are also addressing the `glReadPixels` inversion  on OpenGL by y-inverting the result inside filament. This is done in `Renderer::readPixels`. This may not be ideal, but at least when you call `Renderer::readPixels` and you use it directly as bitmap or send it to a jpeg encoder, you get an image that's not y-inverted with respect to what was on screen.

For textures that will be used only for sampling, there are in fact no issues and you shouldn't use `uvToRenderTargetUV()` for those. This is typically the case for your normal maps, albedo maps, roughness maps, etc...

**The solution**

I think the best solution for your situation is to *physically* invert vertically your image before calling `setImage()` only when using the OpenGL backend. This is not ideal because your code will have to be slightly different when running on an OpenGL backend. You also need to always use `uvToRenderTargetUV()` which will now work on your "inverted" texture as well as on "intermediary" buffers.

You could, of course, do this inversion in the shader, but as you said, you will have to pass a boolean (or the appropriate transform) to apply to the texture coordinates based on whether the source comes from `setImage()` or not.


I hope this helps and I hope this explains why we made the choices we made. Please let us know how it goes.

**Reply by kjw0723** [reply]
@pixelflinger Thank you for such a thorough answer. We originally flipped the image as you recommended, but we're working on finding better solutions. We'll stick to flipping for now, though. Thanks again.
